### PR TITLE
Unit Tests for edge/pkg/edgehub

### LINKED ISSUE #4327: Umbrella issue for test cases enhancement
## Description
We summarized some task items as follows for KubeEdge code quality testing, please comment with one or more items' names if you are interested to join.

### Unit Test
- [x] integrated Codecov into CI/CD -- @gy95 
#### Complete UT cases in the following directory to improve the coverage to at least 80%.
- [ ] kubeedge/cloud/pkg/admissioncontroller @gy95 
- [ ] kubeedge/cloud/pkg/cloudhub @wackxu 
  - kubeedge/cloud/pkg/cloudhub/servers/httpserver/updateconfig_test.go https://github.com/kubeedge/kubeedge/pull/4461 
- [ ] kubeedge/cloud/pkg/cloudstream @wackxu 
- [ ] kubeedge/cloud/pkg/csidriver
  - kubeedge/cloud/pkg/csidriver/uds_test.go https://github.com/kubeedge/kubeedge/pull/4446
- [ ] kubeedge/cloud/pkg/devicecontroller @RyanZhaoXB 
  - kubeedge/cloud/pkg/devicecontroller/manager/devicemodel_test.go https://github.com/kubeedge/kubeedge/pull/4467 
- [ ] kubeedge/cloud/pkg/dynamiccontroller @wackxu 
- [ ] kubeedge/cloud/pkg/edgecontroller @vincentgoat 
- [ ] kubeedge/cloud/pkg/nodeupgraderjobcontroller @gy95 
- [ ] kubeedge/cloud/pkg/router @gy95 
- [x] kubeedge/cloud/pkg/synccontroller @vincentgoat 
- [ ] kubeedge/edge/devicetwin @ksankeerth 
- [x] kubeedge/edge/pkg/edgehub -- @ksankeerth 
- [ ] kubeedge/edge/pkg/eventbus @Nageshbansal 
- [ ] kubeedge/edge/pkg/servicebus @gy95 
- [ ] kubeedge/edge/pkg/edgestream @fisherxu 
- [ ] kubeedge/edge/pkg/metamanager @Shelley-BaoYue 
  - kubeedge/edge/pkg/metamanager/client/nodestatus_test.go https://github.com/kubeedge/kubeedge/pull/4502 
- [ ] kubeedge/edge/pkg/edged @Shelley-BaoYue 
- [ ] kubeedge/keadm/cmd/keadm/app @gy95 
- [ ] kubeedge/pkg/metaserver @Shelley-BaoYue 
- [x] kubeedge/pkg/image @my-git9 
- [ ] kubeedge/pkg/stream @fisherxu 

### E2E Test / Integration Test
- [x] EdgeApplication and NodeGroup features @Congrool 
- [ ] Pods prober check
- [ ] Certificate rotation
- [ ] Meta Server list/watch operations
- [ ] Node Upgrade operations
- [ ] Deploy EdgeMesh for service access in the edge
- [ ] Use DMI with Mapper
- [ ] Device management
- [ ] Message delivered by Servicebus/Eventbus/REST
- [ ] Reliable message management in the module synccontroller
- [ ] Debugging operations of logs/exec in the module cloudstream/edgestream
- [ ] Metrics server

### Conformance Testing
- [x] Integrating conformance tests framework and the base conformance tests @wackxu @vincentgoat 
- [ ] Adding conformance tests of Devices and Rule/RuleEndpoint

### PR DESCRIPTION
**Pull request by ksankeerth**
**What type of PR is this?**

/kind test

**What this PR does / why we need it**:
This is PR add a few more tests to edge/pkg/edgehub

**Which issue(s) this PR fixes**:

Fixes #4327  (only one UT for edge/pkg/edgehub)

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
NONE

```release-note

```

### CODE REVIEW

File: edge/pkg/edgehub/edgehub_test.go [hunk 1]
```diff
@@ -1,95 +1,302 @@
...
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if got := tt.eh.Enable(); got != tt.want {
+				t.Errorf("EdgeHub.Enable() returned expected results. got = %v, want = %v", got, tt.want)
+			}
+		})
+	}
+}
+
+// TODO: eh.Start will wait until we call beehiveContext.Cancel(). When running `make test` tests, it seems that
```
Comment (ksankeerth) [inline edge/pkg/edgehub/edgehub_test.go:211]: Hi Team,
I'm was not able to add test Start() function. 

Inside the `Start()`, We wait for `beehiveContext.Done() channel.  The context done is channel is a global variable (initialised by init() function). It always blocks the `testing.Run()` forever. (I think this is due to nil context.Done() channel)

If I run only one test(There are two testcases 1. Websocket  2.Quic), then it'll work.  If I run the both test cases, then the sample problem will happen.

**Comment by ksankeerth** [comment]
I've fixed lint issues.

**Comment by ksankeerth** [comment]
Hi Team,
Need some help to identity the reason for CI failure.  I checked the CI workflow logs, [CI logs](https://github.com/kubeedge/kubeedge/actions/runs/3368305828/jobs/5600610941#step:8:1346).

One e2e test fails after getting timeout. The test is delete statefulSet pod multipleTimes and waiting to k8s to create a new pod. (timeout is 120s)

Not sure, how the newly added unit tests affect integration tests.  Do we need to increase timeout? In other places, I can see that we are waiting for 240s before fails.

eg : 
```go
			utils.WaitforPodsRunning(ctx.Cfg.KubeConfigPath, *podList, 240*time.Second)

			deletePodName := fmt.Sprintf("%s-1", UID)
			for i := 0; i < 5; i++ {
				By(fmt.Sprintf("delete pod %s", deletePodName))
				err = utils.DeletePod(clientSet, deletePodName, "default")
				Expect(err).To(BeNil())

				By(fmt.Sprintf("wait for pod %s running again", fmt.Sprintf("%s-1", UID)))
				err = wait.Poll(5*time.Second, 120*time.Second, func() (bool, error) {
					pod, err := clientSet.CoreV1().Pods("default").Get(context.TODO(), deletePodName, metav1.GetOptions{})
					if err != nil {
						return false, err
					}
					if pod.Status.Phase == corev1.PodRunning {
						return true, nil
					}
					return false, nil
				})
				Expect(err).To(BeNil())
```

```console
 Nov  2 04:07:39.022: Info: TestCase:      Delete statefulSet pod multi times     Status=FAILED
  << End Captured GinkgoWriter Output

  Expected
      <*errors.errorString | 0xc000063c10>: {
          s: "timed out waiting for the condition",
      }
  to be nil
  In [It] at: /home/runner/work/kubeedge/kubeedge/tests/e2e/deployment/deployment_test.go:304
```

File: edge/pkg/edgehub/edgehub_test.go (hunk 1 above)
Comment (vincentgoat) [reply edge/pkg/edgehub/edgehub_test.go:206]: Thanks!
I'd like to double check this block.

Comment (vincentgoat) [reply edge/pkg/edgehub/edgehub_test.go:206]: @ksankeerth I tested on my local host, we can try to new a mock client same as here https://github.com/kubeedge/kubeedge/blob/master/edge/pkg/edgehub/process_test.go#L126.
That would be well to do the unit test.

Comment (ksankeerth) [reply edge/pkg/edgehub/edgehub_test.go:206]: Thanks for the suggestion. I'll update the PR  👍

Comment (ksankeerth) [reply edge/pkg/edgehub/edgehub_test.go]: I wrote tests for `EdgeHub.Start()` and seems to be working. Please check. Thanks

**Comment by vincentgoat** [comment]
@ksankeerth Sorry for the late reply, thanks for your work on debugging. I agree with your suggestion.
We can improve it later if necessary.

**Comment by ksankeerth** [comment]
Hi @vincentgoat,
I'm trying to fix CI-Fuzz failures.  The below are error logs.

```console
Running go-fuzz -tags gofuzz -func FuzzdealTwinActions -o fuzz_deal_twin_actions.a github.com/kubeedge/kubeedge/edge/pkg/devicetwin/dtmanager
# golang.org/x/sys/unix
vendor/golang.org/x/sys/unix/syscall.go:83:28: unsafe.Slice requires go1.17 or later (-lang was set to go1.16; check go.mod)
vendor/golang.org/x/sys/unix/syscall_linux.go:2256:21: unsafe.Slice requires go1.17 or later (-lang was set to go1.16; check go.mod)
vendor/golang.org/x/sys/unix/syscall_unix.go:118:19: unsafe.Slice requires go1.17 or later (-lang was set to go1.16; check go.mod)
vendor/golang.org/x/sys/unix/sysvshm_unix.go:33:19: unsafe.Slice requires go1.17 or later (-lang was set to go1.16; check go.mod)
2022/11/21 09:49:59 failed to build packages:exit status 2
2022-11-21 09:50:02,[577](https://github.com/kubeedge/kubeedge/actions/runs/3462512911/jobs/5885834108#step:4:578) - root - ERROR - Building fuzzers failed.
2022-11-21 09:50:02,[578](https://github.com/kubeedge/kubeedge/actions/runs/3462512911/jobs/5885834108#step:4:579) - root - ERROR - Error building fuzzers for (commit: 6c263fe9da3daa998f4ecc8a4d3bab0bb6e6a0e2, pr_ref: refs/pull/4360/merge).
```
https://github.com/kubeedge/kubeedge/actions/runs/3462512911/jobs/5885834108#step:4:572

It seems that fuzz build failed due to Go version in kubeedge/kubeedge/edge/pkg/devicetwin/dtmanager. Not sure, It's relevant to this PR. I don't know much about fuzz test. Any pointers to start fixing these errors?

**Comment by vincentgoat** [comment]
Hi @ksankeerth,
Please forget the Fuzzing error, it is a Fuzzing suite integration error and has been solved now.
FYI: https://github.com/kubeedge/kubeedge/issues/4426
I will re-run the CIFUZZ to solve this error.

File: edge/pkg/edgehub/certificate/certmanager_test.go [hunk 2]
```diff
@@ -0,0 +1,65 @@
+package certificate
```
Comment (vincentgoat) [inline edge/pkg/edgehub/certificate/certmanager_test.go:1]: Copyright should be added in a new file, and the below files also should add.

Comment (ksankeerth) [reply edge/pkg/edgehub/certificate/certmanager_test.go:17]: done. Added copy right headers to all the newly added and modified files.

File: edge/pkg/edgehub/certificate/certmanager_test.go [hunk 3]
```diff
@@ -0,0 +1,81 @@
+/*
+Copyright 2019 The KubeEdge Authors.
```
Comment (vincentgoat) [inline edge/pkg/edgehub/certificate/certmanager_test.go:2]: The copy right date should be 2022.

Comment (ksankeerth) [reply edge/pkg/edgehub/certificate/certmanager_test.go]: sorry for the mistake. Done

**Review by fisherxu — COMMENTED** [review]
/lgtm
/approve
